File: client-library/src/api-client.ts
/**
 * Koro i18n API Client
 * 
 * Handles communication with the koro-i18n platform API.
 * Used by the CLI for push/pull operations.
 */

// ============================================================================
// Types
// ============================================================================

export interface SourceKeyOperation {
  op: 'add' | 'update' | 'delete';
  filename: string;
  key: string;
  value?: string;
  hash?: string;
}

export interface SyncResult {
  success: boolean;
  results?: {
    added: number;
    updated: number;
    deleted: number;
    errors: string[];
  };
  error?: string;
}

export interface ImportTranslation {
  language: string;
  filename: string;
  key: string;
  value: string;
  hash?: string;
}

export interface ImportResult {
  success: boolean;
  results?: {
    imported: number;
    skipped: number;
    errors: string[];
  };
  error?: string;
}

export interface HashManifest {
  success: boolean;
  manifest: Record<string, Record<string, string>>; // filename -> key -> hash
  totalKeys: number;
}

export interface Translation {
  language: string;
  filename: string;
  key: string;
  value: string;
  status: 'draft' | 'approved';
  sourceHash?: string;
}

export interface PullResult {
  success: boolean;
  translations: Translation[];
  error?: string;
}

// ============================================================================
// API Client
// ============================================================================

export class KoroApiClient {
  private baseUrl: string;
  private projectName: string;
  private token: string | null;

  constructor(options: {
    baseUrl?: string;
    projectName: string;
    token?: string;
  }) {
    this.baseUrl = options.baseUrl || process.env.KORO_API_URL || 'https://koro.f3liz.workers.dev';
    this.projectName = options.projectName;
    this.token = options.token || this.getToken();
  }

  /**
   * Get authentication token from environment
   * Supports GitHub Actions OIDC and JWT token
   */
  private getToken(): string | null {
    // GitHub Actions OIDC - fetch token from OIDC provider
    if (process.env.ACTIONS_ID_TOKEN_REQUEST_TOKEN && process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
      // Token will be fetched dynamically in getHeaders()
      return null;
    }
    
    // Environment variable
    return process.env.KORO_TOKEN || null;
  }

  /**
   * Get authorization headers
   */
  private async getHeaders(): Promise<Record<string, string>> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    // GitHub Actions OIDC
    if (process.env.ACTIONS_ID_TOKEN_REQUEST_TOKEN && process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
      try {
        const response = await fetch(
          `${process.env.ACTIONS_ID_TOKEN_REQUEST_URL}&audience=${this.baseUrl}`,
          {
            headers: {
              Authorization: `bearer ${process.env.ACTIONS_ID_TOKEN_REQUEST_TOKEN}`,
            },
          }
        );
        const data = await response.json() as { value: string };
        headers['Authorization'] = `Bearer ${data.value}`;
      } catch (e) {
        console.error('Failed to fetch OIDC token:', e);
      }
    } else if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }

    return headers;
  }

  /**
   * Get hash manifest for client-side diffing
   */
  async getHashManifest(): Promise<HashManifest> {
    const headers = await this.getHeaders();
    const response = await fetch(
      `${this.baseUrl}/api/projects/${this.projectName}/source-keys/hash`,
      { headers }
    );
    
    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Failed to get hash manifest: ${error}`);
    }
    
    return response.json() as Promise<HashManifest>;
  }

  /**
   * Sync source keys (push)
   */
  async syncSourceKeys(
    operations: SourceKeyOperation[],
    commitSha?: string
  ): Promise<SyncResult> {
    const headers = await this.getHeaders();
    const response = await fetch(
      `${this.baseUrl}/api/projects/${this.projectName}/source-keys/sync`,
      {
        method: 'POST',
        headers,
        body: JSON.stringify({ commitSha, operations }),
      }
    );

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    return response.json() as Promise<SyncResult>;
  }

  /**
   * Import existing translations
   */
  async importTranslations(
    translations: ImportTranslation[],
    autoApprove: boolean = false
  ): Promise<ImportResult> {
    const headers = await this.getHeaders();
    const response = await fetch(
      `${this.baseUrl}/api/projects/${this.projectName}/source-keys/import-translations`,
      {
        method: 'POST',
        headers,
        body: JSON.stringify({ translations, autoApprove }),
      }
    );

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    return response.json() as Promise<ImportResult>;
  }

  /**
   * Pull approved translations
   */
  async pullTranslations(options?: {
    language?: string;
    status?: 'draft' | 'approved';
  }): Promise<PullResult> {
    const headers = await this.getHeaders();
    const params = new URLSearchParams();
    
    if (options?.language) params.set('language', options.language);
    if (options?.status) params.set('status', options.status);
    
    const url = `${this.baseUrl}/api/projects/${this.projectName}/apply/export${
      params.toString() ? `?${params}` : ''
    }`;
    
    const response = await fetch(url, { headers });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, translations: [], error };
    }

    const data = await response.json() as { translations: Translation[] };
    return { success: true, translations: data.translations };
  }
}

// ============================================================================
// Diff Utilities
// ============================================================================

export interface LocalKey {
  filename: string;
  key: string;
  value: string;
  hash: string;
}

/**
 * Compute diff between server and local keys
 */
export function computeDiff(
  serverManifest: Record<string, Record<string, string>>,
  localKeys: LocalKey[]
): SourceKeyOperation[] {
  const operations: SourceKeyOperation[] = [];
  const localMap = new Map<string, LocalKey>();

  // Build local map: "filename:key" -> LocalKey
  for (const key of localKeys) {
    localMap.set(`${key.filename}:${key.key}`, key);
  }

  // Build server set
  const serverSet = new Set<string>();
  for (const [filename, keys] of Object.entries(serverManifest)) {
    for (const key of Object.keys(keys)) {
      serverSet.add(`${filename}:${key}`);
    }
  }

  // Find adds and updates
  for (const local of localKeys) {
    const id = `${local.filename}:${local.key}`;
    const serverHash = serverManifest[local.filename]?.[local.key];

    if (!serverHash) {
      // New key
      operations.push({
        op: 'add',
        filename: local.filename,
        key: local.key,
        value: local.value,
        hash: local.hash,
      });
    } else if (serverHash !== local.hash) {
      // Updated key
      operations.push({
        op: 'update',
        filename: local.filename,
        key: local.key,
        value: local.value,
        hash: local.hash,
      });
    }
  }

  // Find deletes
  for (const serverId of serverSet) {
    if (!localMap.has(serverId)) {
      const [filename, key] = serverId.split(':');
      operations.push({
        op: 'delete',
        filename,
        key,
      });
    }
  }

  return operations;
}

File: client-library/lib/koro_i18n/api_client.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

module KoroI18n
  # API Client for communicating with the Koro i18n platform
  class ApiClient
    attr_reader :base_url, :project_name

    def initialize(project_name:, base_url: nil, token: nil)
      @base_url = base_url || ENV["KORO_API_URL"] || "https://koro.f3liz.workers.dev"
      @project_name = project_name
      @token = token || get_token
    end

    # Get hash manifest for client-side diffing
    def get_hash_manifest
      uri = URI("#{@base_url}/api/projects/#{@project_name}/source-keys/hash")
      response = make_request(uri, :get)

      raise Error, "Failed to get hash manifest: #{response.body}" unless response.is_a?(Net::HTTPSuccess)

      data = JSON.parse(response.body)
      {
        success: true,
        manifest: data["manifest"] || {},
        total_keys: data["totalKeys"] || 0
      }
    end

    # Sync source keys (push)
    def sync_source_keys(operations, commit_sha = nil)
      uri = URI("#{@base_url}/api/projects/#{@project_name}/source-keys/sync")
      body = { commitSha: commit_sha, operations: operations.map(&:to_h) }

      response = make_request(uri, :post, body)

      unless response.is_a?(Net::HTTPSuccess)
        return { success: false, error: response.body }
      end

      data = JSON.parse(response.body)
      {
        success: true,
        results: {
          added: data.dig("results", "added") || 0,
          updated: data.dig("results", "updated") || 0,
          deleted: data.dig("results", "deleted") || 0,
          errors: data.dig("results", "errors") || []
        }
      }
    end

    # Import existing translations
    def import_translations(translations, auto_approve: false)
      uri = URI("#{@base_url}/api/projects/#{@project_name}/source-keys/import-translations")
      body = {
        translations: translations.map do |t|
          {
            language: t[:language],
            filename: t[:filename],
            key: t[:key],
            value: t[:value],
            hash: t[:hash]
          }
        end,
        autoApprove: auto_approve
      }

      response = make_request(uri, :post, body)

      unless response.is_a?(Net::HTTPSuccess)
        return { success: false, error: response.body }
      end

      data = JSON.parse(response.body)
      {
        success: true,
        results: {
          imported: data.dig("results", "imported") || 0,
          skipped: data.dig("results", "skipped") || 0,
          errors: data.dig("results", "errors") || []
        }
      }
    end

    # Pull approved translations
    def pull_translations(language: nil, status: nil)
      uri = URI("#{@base_url}/api/projects/#{@project_name}/apply/export")

      params = []
      params << "language=#{language}" if language
      params << "status=#{status}" if status
      uri.query = params.join("&") unless params.empty?

      response = make_request(uri, :get)

      unless response.is_a?(Net::HTTPSuccess)
        return { success: false, translations: [], error: response.body }
      end

      data = JSON.parse(response.body)
      translations = (data["translations"] || []).map do |t|
        Translation.new(
          language: t["language"],
          filename: t["filename"],
          key: t["key"],
          value: t["value"],
          status: t["status"],
          source_hash: t["sourceHash"]
        )
      end

      { success: true, translations: translations }
    end

    private

    def get_token
      # GitHub Actions OIDC
      if ENV["ACTIONS_ID_TOKEN_REQUEST_TOKEN"] && ENV["ACTIONS_ID_TOKEN_REQUEST_URL"]
        # Token will be fetched dynamically in get_headers
        return nil
      end

      # Environment variable
      ENV["KORO_TOKEN"]
    end

    def get_headers
      headers = { "Content-Type" => "application/json" }

      # GitHub Actions OIDC
      if ENV["ACTIONS_ID_TOKEN_REQUEST_TOKEN"] && ENV["ACTIONS_ID_TOKEN_REQUEST_URL"]
        begin
          oidc_uri = URI("#{ENV['ACTIONS_ID_TOKEN_REQUEST_URL']}&audience=#{@base_url}")
          http = Net::HTTP.new(oidc_uri.host, oidc_uri.port)
          http.use_ssl = true

          request = Net::HTTP::Get.new(oidc_uri)
          request["Authorization"] = "bearer #{ENV['ACTIONS_ID_TOKEN_REQUEST_TOKEN']}"

          response = http.request(request)
          if response.is_a?(Net::HTTPSuccess)
            data = JSON.parse(response.body)
            headers["Authorization"] = "Bearer #{data['value']}"
          end
        rescue StandardError => e
          $stderr.puts "Failed to fetch OIDC token: #{e.message}"
        end
      elsif @token
        headers["Authorization"] = "Bearer #{@token}"
      end

      headers
    end

    def make_request(uri, method, body = nil)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"

      request = case method
                when :get
                  Net::HTTP::Get.new(uri)
                when :post
                  Net::HTTP::Post.new(uri)
                end

      get_headers.each { |k, v| request[k] = v }
      request.body = JSON.generate(body) if body

      http.request(request)
    end
  end
end
